### PR TITLE
Fall back to heartbeat time when converted span snapshots

### PR DESCRIPTION
**Pull request by bidetofevil**
## Goal

Use the last heartbeat time of the session span if the end time was missing, like if the session itself wasn't terminated properly, like during a native crash or a user kill while the app is in background.

Previously we were assuming that there's always an end time, and we even tested for that case too, which yield and end of 0 in those cases. The backend is able to work with that in most case, as it just figures it out by looking at other timestamps. So while customers don't see this problem for the most part, like for sessions associated with native crashes, the session span is wrong. I was also seeing some 0-length sessions for background kills where the session span contains a heartbeat time, so it's better than we do this population properly, and test it properly as well.

## Testing
Modify and extend unit test on the `EmbraceDeliveryService`. Tweaking some testing infra to make this easier too.

### CODE REVIEW

**Comment by bidetofevil** [comment]
* **#1229** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1229?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bidetofevil and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/comms/delivery/EmbraceDeliveryService.kt [hunk 1]
```diff
@@ -191,8 +194,18 @@ internal class EmbraceDeliveryService(
...
 
-    private fun getFailedSpanEndTimeMs(envelope: Envelope<SessionPayload>) =
-        envelope.getSessionSpan()?.endTimeNanos?.nanosToMillis() ?: -1
+    /**
+     * To approximate the time of any snapshot to be converted into a failed span, we look to the session span of the payload and take
+     * either the end time or the last heartbeat time, whichever exists and is later. If the session span itself is a snapshot, it will
+     * not have an end time, in which case it will fall back to the last heartbeat time. If either exists, it means we can't find a better
+     * time, so we just leave it at 0.
+     */
+    private fun getFailedSpanEndTimeMs(envelope: Envelope<SessionPayload>): Long {
```
Comment (fractalwrench) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/comms/delivery/EmbraceDeliveryService.kt:203]: I found this a bit hard to follow, would suggest breaking it down with local variables. E.g.

```
    private fun getFailedSpanEndTimeMs(envelope: Envelope<SessionPayload>): Long {
        val sessionSpan = envelope.getSessionSpan()
        val endTime = sessionSpan?.endTimeNanos ?: 0L
        val heartbeatTime = sessionSpan?.attributes?.findAttributeValue(embHeartbeatTimeUnixNano.attributeKey.key)?.toLongOrNull() ?: 0L
        return max(endTime, heartbeatTime).nanosToMillis()
    }
```

**Review by fractalwrench — APPROVED** [review]
LGTM

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/comms/delivery/EmbraceDeliveryService.kt (hunk 1 above)
Comment (bidetofevil) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/comms/delivery/EmbraceDeliveryService.kt:203]: yeah good idea. I'll do that!

I also want to pull this out of this class because it's not really a delivery concern, but I didn't want to do that now, or when I rejiggered the logic initially. I should do it soon....

**Comment by bidetofevil** [comment]
### Merge activity

* **Aug 12, 12:20 PM EDT**: @bidetofevil merged this pull request with [Graphite](https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1229).
